[googledatastore] Add a Google Datastore client

stfeng opened the pull request:
Add a DB client for [Google Cloud Datastore](https://cloud.google.com/datastore/docs/concepts/overview?hl=en)

This Pull Request includes implementation of all CRUD operations (except scan), and integration into the rest of YCSB (pom.xml changes, etc).

cmccoy commented on googledatastore/src/main/java/com/yahoo/ycsb/db/GoogleDatastoreClient.java:
```diff
@@ -0,0 +1,334 @@
+    try {
+      // Setup the connection to Google Cloud Datastore with the credentials
+      // obtained from the configure.
+      DatastoreOptions.Builder options = new DatastoreOptions.Builder();
```
Can this use [application default credentials](https://developers.google.com/identity/protocols/application-default-credentials#callingjava)?

stfeng replied:
I think if we are running on GCE VM, we could use application default credentials if we create the VM with a special flag. ([here](https://cloud.google.com/datastore/docs/getstarted/start_java/) is the datastore doc about this.)

However, if we want this tool to be able to run from somewhere else other than a GCE VM, then we need to take a more generic approach to setup and use credentials (as shown in step 2 in the doc linked above), which is the approach I'm taking here.

busbey replied:
Please file a follow-on issue to enable falling back to try default creds when on GCE.

stfeng replied:
Done. Filed #510

k1xme replied:
@stfeng can we somehow support both ADC and credential file authentication in there? Like adding a flag to switch between these 2 modes. This will be very handy.

kruthar commented on googledatastore/src/main/java/com/yahoo/ycsb/db/GoogleDatastoreClient.java:
```diff
@@ -0,0 +1,334 @@
+    if (response.getFoundCount() == 0) {
+      return new Status("ERROR-404", "Not Found, key is: " + key);
+    }
+
```
Is there possibility of response.getFoundCount() > 1? This would be an unexpected result, but would get be marked as key not found.

stfeng replied:
Good point! I've added a check for getFoundCount()>1 and return unexpected state from there. Since we only requested to lookup for one key, so we are only expecting one entity back.

kruthar commented:
The documentation in googledatastore.properties is great, but looking at other bindings the convention seems to be to document binding specific configuration in the bindings README.

kruthar commented:
Also, can you add checkstyle enforcement to your pom.xml and fix any issues that come up in the build? See here: https://github.com/brianfrankcooper/YCSB/blob/master/accumulo/pom.xml#L67-L90

kruthar commented on googledatastore/conf/googledatastore.properties:
```diff
@@ -0,0 +1,92 @@
+#
+# googledatastore.datasetId=<string id of your dataset>
+# googledatastore.privateKeyFile=<full path to your private key file>
+# googledatastore.serviceAccountEmail=<Your service account email>
```
GoogleDatastoreClient.java is actually expecting ```googledatastore.serviceAccount``` as the property name.

RogerKirby replied:
What exactly is the googledatastore.datasetId=<string id of your dataset> referring to? I read on another forum that it is Project ID but when I update the properties file with the Project ID the following error is returned:

site.ycsb.DBException: Required property "datasetId" missing.
        at site.ycsb.db.GoogleDatastoreClient.init(GoogleDatastoreClient.java:113)
        at site.ycsb.DBWrapper.init(DBWrapper.java:86)
        at site.ycsb.ClientThread.run(ClientThread.java:91)
        at java.base/java.lang.Thread.run(Thread.java:832)

I've tried using the Project Number but to no avail.

kruthar commented:
When testing this I received this:

```sh
com.yahoo.ycsb.DBException: I/O error connecting to the datastore: toDerInputStream rejects tag type 123
	at com.yahoo.ycsb.db.GoogleDatastoreClient.init(GoogleDatastoreClient.java:177)
	at com.yahoo.ycsb.DBWrapper.init(DBWrapper.java:63)
	at com.yahoo.ycsb.ClientThread.run(Client.java:289)
```

The fix was to simply use a P12 private key file instead of a Json key file. Is that a know limitation, and hence should be documented in the README or somewhere? Or was that just an issue with my particular setup?

stfeng commented on googledatastore/conf/googledatastore.properties:
```diff
@@ -0,0 +1,92 @@
+#
+# googledatastore.datasetId=<string id of your dataset>
+# googledatastore.privateKeyFile=<full path to your private key file>
+# googledatastore.serviceAccountEmail=<Your service account email>
```
Yep, I agree this variable name should be serviceAccountEmail. And thanks for the catch!

stfeng commented:
@kruthar, Thanks for the great code review! I've incorporated all your comments into my code. Here are  answers to your questions:

1. Yes the Google Cloud Datastore needs a p12 private key file to authenticate. I added a note to README.md to clarify this.

2. Checkstyle: I've added the checkstyle plugin to pom.xml and fixed all issues (except one issue about TODO comment, which I'd like to keep there until the work is actually done. :-)

3. Moved documentation from .properties file to READMD.md

PTAL, thanks!

kruthar commented on googledatastore/src/main/java/com/yahoo/ycsb/db/GoogleDatastoreClient.java:
```diff
@@ -0,0 +1,339 @@
+          serviceAccountEmail, privateKeyFile);
+      logger.info("Using JWT Service Account credential.");
+      logger.info("DatasetID: " + datasetId + "\nService Account Email: " +
+          serviceAccountEmail + "\nPrivate Key File Path: " + privateKeyFile);
```
Log statements should be greppable single line messages, this could just be broken up into three logger.info statements instead of using newlines.

kruthar commented on googledatastore/src/main/java/com/yahoo/ycsb/db/GoogleDatastoreClient.java:
```diff
@@ -0,0 +1,339 @@
+    }
+
+    logger.info("Datastore client instance created:\n" +
+        datastore.toString());
```
Same here, this could just be a single log line.

kruthar commented on googledatastore/src/main/java/com/yahoo/ycsb/db/GoogleDatastoreClient.java:
```diff
@@ -0,0 +1,339 @@
+    // Note above, datastore lookupRequest always reads the entire entity, it
+    // does not support reading a subset of "fields" (properties) of an entity.
+
+    logger.debug("Built lookup request as:\n" + lookupRequest.toString());
```
And once more, remove the newline so this is a single line in the log output.

kruthar commented:
One last set of inline comments about multiline log statements. Otherwise I built this module successfully, and ran some simple load/read/update tests and things look good!

stfeng commented:
Done. Incorporated all comments, changed from newlines to one line output in the debug log. 

Thank you for the CR and taking the extra step to build, test and validate this module. Much appreciated!

stfeng commented:
@kruthar @cmccoy 

Is this ready to merge? Thanks!

kruthar commented:
Looks good to me.

busbey commented:
Could you update the last two commits to have a message that starts with the module that's changed? Maybe squash commits as well?

busbey commented on googledatastore/README.md:
```diff
@@ -0,0 +1,91 @@
+<!--
+Copyright (c) 2010 Yahoo! Inc., 2012 - 2015 YCSB contributors.
```
Any reason this isn't just (c) 2015 contributors?

stfeng replied:
Done. Changed this and everywhere else to (c) 2015 YCSB contributors.

busbey commented on googledatastore/pom.xml:
```diff
@@ -0,0 +1,73 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!-- 
+Copyright (c) 2012 - 2015 YCSB contributors. All rights reserved.
```
Why 2012-2015?

busbey commented on googledatastore/pom.xml:
```diff
@@ -0,0 +1,73 @@
+      <plugin>
+        <groupId>org.apache.maven.plugins</groupId>
+        <artifactId>maven-checkstyle-plugin</artifactId>
+        <version>2.15</version>
```
Do we not set the checkstyle version in the root pom?

kruthar replied:
There is checkstyle validation in the root pom, but the version is not parameterized to use in the modules. Adding the checkstyle enforcement in each module is really just a way to incrementally add verification to the modules, once they all have it and the entire project is cleaned up, it will probably be best to just at failOnViolation to the root and remove all enforcements at the module level as it is largely repetitious.

stfeng replied:
Right, when @kruthar asked me to add the checkstyle plugin into this binding specific pom.xml file, I had the same question, then I realized the parent/root checkstyle does not fail the build when there is violation, and there doesn't appear to be a way to parameterize this version of checkstyle plug-in to make it behave differently. 

This is the same way of configuration in other DB bindings, such as Accumulo. 

What's important here though is we are still sharing exactly the same set of checkstyle rules and configurations (../checkstyle.xml) with the rest of the YCSB code base.

busbey commented on googledatastore/src/main/java/com/yahoo/ycsb/db/GoogleDatastoreClient.java:
```diff
@@ -0,0 +1,339 @@
+/*
+ * Copyright 2015 Google Inc. All Rights Reserved.
```
Does Google need to retain a by name copyright here?

busbey commented on googledatastore/src/main/java/com/yahoo/ycsb/db/GoogleDatastoreClient.java:
```diff
@@ -0,0 +1,339 @@
+
+/**
+ * Google Cloud Datastore Client for YCSB.
+ * Author: stfeng@
```
No author tags please.

busbey commented on googledatastore/src/main/java/com/yahoo/ycsb/db/GoogleDatastoreClient.java:
```diff
@@ -0,0 +1,339 @@
+
+  // Read consistency defaults to "eventual" (this is the same as other
+  // DB client, such as DynamoDB). User can override this via configure.
+  private ReadConsistency readConsistency = ReadConsistency.EVENTUAL;
```
Apologies for not having made a guide for new bindings yet. YCSB clients should default to the strictest consistency available and document consistency tuning and tradeoffs in their readme.

stfeng replied:
I'm totally fine setting this default to STRONG here. Happy to do that.

Meanwhile, I also think we should set other DB clients default to conform with this guideline too. I've filed issue #509 to set the DynamoDB client's read consistency default level to STRONG so it becomes comparable to the rest of the YCSB clients at default setting.

busbey commented on googledatastore/src/main/java/com/yahoo/ycsb/db/GoogleDatastoreClient.java:
```diff
@@ -0,0 +1,339 @@
+          options.credential(credential).dataset(datasetId).build());
+
+    } catch (GeneralSecurityException exception) {
+      throw new DBException("Security error connecting to the datastore: " +
```
Can we wrap causes in all these cases instead of just grabbing the message?

stfeng replied:
Done.

busbey replied:
Sorry for not making this clear earlier, by "wrap" I mean to use the [two argument constructor that takes a Throwable cause](https://github.com/brianfrankcooper/YCSB/blob/master/core/src/main/java/com/yahoo/ycsb/DBException.java#L40).

e.g.

```java
  throw new DBException("Security error connecting to the datastore", exception)
```

stfeng replied:
Done.

busbey replied:
should be exception rather than exception.getCause

stfeng replied:
Agree! Done.

busbey commented on googledatastore/src/main/java/com/yahoo/ycsb/db/package-info.java:
```diff
@@ -0,0 +1,22 @@
+/**
+ * Copyright (c) 2015 Google Inc. All rights reserved.
```
Same coment as above re: Google as named copyright rather than the general "YCSB contributors"

stfeng commented:
@busbey, I've incorporated all your comments and pushed the changes. I've modified the commit message to all start with [googledatastore] and squashed a couple of commits. 

PTAL. Thanks.

stfeng commented:
@busbey Could you please take a look and let me know if there is anything else I need to do to get this merged? Thanks.

busbey commented:
the DBException comment above is the only issue I see.

stfeng commented:
Done. Changed the DBException() to take the cause. PTAL. Thanks.

stfeng commented:
Pushed the last commit (and squashed it into the previous one) that addressed the last comment. Thanks for all the reviews.

busbey commented:
thanks for the contribution!

k1xme commented on googledatastore/src/main/java/com/yahoo/ycsb/db/GoogleDatastoreClient.java:
```diff
@@ -0,0 +1,338 @@
+  }
+
+  @Override
+  public Status scan(String table, String startkey, int recordcount,
```
Why don't we implement this method? Does YCSB not call it?
